pkg/controller/usersignup/usersignup_controller.go

```diff
@@ -0,0 +1,187 @@
+		predicate.GenerationChangedPredicate{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileUserSignup{}
+
+// ReconcileUserSignup reconciles a UserSignup object
+type ReconcileUserSignup struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	clientset kubernetes.Interface
+}
+
+// Reconcile reads that state of the cluster for a UserSignup object and makes changes based on the state read
+// and what is in the UserSignup.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileUserSignup) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling UserSignup")
+
+	// Fetch the UserSignup instance
+	instance := &toolchainv1alpha1.UserSignup{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Read the status conditions and assign some variables
+	provisioning := condition.IsTrue(instance.Status.Conditions, toolchainv1alpha1.UserSignupProvisioning)
+	completed := condition.IsTrue(instance.Status.Conditions, toolchainv1alpha1.UserSignupComplete)
+
+	// If the status is not yet provisioning or completed, check for automatic or manual approval
+	if !provisioning && !completed {
+		userApprovalPolicy, err := r.ReadUserApprovalPolicyConfig()
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// If the signup has been explicitly approved (by an admin), or the user approval policy is set to automatic,
+		// then proceed with the signup
+		if instance.Spec.Approved || userApprovalPolicy == config.UserApprovalPolicyAutomatic {
+			var targetCluster string
+
+			// If a target cluster hasn't been selected, select one from the members
+			if instance.Spec.TargetCluster != "" {
+				targetCluster = instance.Spec.TargetCluster
+			} else {
+				// Automatic cluster selection
+				members := cluster.GetMemberClusters()
+				if len(members) > 0 {
+					targetCluster = members[0].Name
+				}
+
+    			// TODO how do we handle the situation where no members are available?
+
+			}
+
+			// Provision the MasterUserRecord
+			userAccounts := []toolchainv1alpha1.UserAccountEmbedded{
+				{
+					TargetCluster: targetCluster,
```

Comment:
@sbryzak here we just set `TargetCluster` only .. how about setting  `UserAccountSpec` ?? [link](https://github.com/codeready-toolchain/api/blob/master/pkg/apis/toolchain/v1alpha1/masteruserrecord_types.go#L52)

I think that first we should have clear picture on how UserAccount CR will be created on member_cluster so that we can decide above changes correctly.  cc @alexeykazakov

Reply:
As we discussed, this will be handled by the `MasterUserRecordController`.

Reply:
I don't think so @sbryzak. The MUR.spec should be completely set by the `UserSignupController`. Remember the reason why we introduced the `UserSignup` - the controller should not modify the spec of the primary resources that watches, so it cannot be handled by `MasterUserRecordController`. 
In other words, the `UserSignupController` should create the MUR with the whole spec and with all information as the desired state (including template versions etc...)

Reply:
@MatousJobanek Nurali and I had a discussion about this already, and I believe that we agreed if the spec wasn't specified, then the `MasterUserRecordController` would create the `UserAccount` resource with sensible default values.  It would not modify the spec of the `MasterUserRecord` resource.  We can certainly have `UserSignupController` populate the MUR with all the values, however that's just moving the burden of doing this just one step higher, the values still need to come from somewhere.

Reply:
@sbryzak
1. we need to have the MUR complete with all information there (including template versions, tiers, limits, etc...)
2. we cannot do it from MUR controller

who else could do that than `UserSignupController`? As I see it, `UserSignupController` should create MUR with spec having all desired values.

Reply:
check this part of the design doc: https://docs.google.com/document/d/19UzaXVqiOo9JqDYz3qUKguJwW-7zceA4UtTr358R_CA/edit?ts=5cef47fb#heading=h.2qyu2s6p5i8c
to have it working, we need to have the values set in the MUR.spec

>  We can certainly have UserSignupController populate the MUR with all the values, however that's just moving the burden of doing this just one step higher, the values still need to come from somewhere.

yes, that's exactly right. So why not have the burden of doing this in the best place it can be

Reply:
@MatousJobanek I will need all of the values from which to populate the `MasterUserRecord` then.  BTW the document link doesn't seem to link to a particular section, did you mean to link to `MasterUserRecord` under "Custom Resources"?

Reply:
> I will need all of the values from which to populate the MasterUserRecord then.

yup - the values from NsTemplateTier which hasn't been implemented yet. Just add a TODO statement there to not forget (maybe also with a reference to Jira)

> BTW the document link doesn't seem to link to a particular section, did you mean to link to MasterUserRecord under "Custom Resources"?

No, the link refers to the heading "Updating namespace template to a newer revision"

Reply:
I've added the TODO statement, with a reference to CRT-74 (I'm not sure if there is a better fit).  So my understanding is that we will now address this as a separate concern, in a different PR.  If you can confirm that CRT-74 is the most appropriate issue for this, I'll update the issue description with some notes on what we need to change in `UserSignupController`.

Reply:
CRT-74 is more for creation of the namespaces on the member side.
To be honest, I don't think that we have any issue for populating MUR with the information from NsTemplateTier and also not for NsTemplateTier management itself. Feel free to create a new one.